src/server.rs

```diff
@@ -356,64 +401,75 @@ impl LanguageServer for LspServer {
         params: lsp::DidChangeTextDocumentParams,
     ) -> () {
         let key = params.text_document.uri;
-        let mut store = match self.store.lock() {
-            Ok(value) => value,
-            Err(err) => {
-                log::warn!(
-                    "Could not acquire store lock. Error: {}",
-                    err
-                );
-                return;
-            }
-        };
-        let mut contents = if let Some(contents) = store.get(&key) {
-            Cow::Borrowed(contents)
-        } else {
-            log::error!(
+        let contents = {
+            let mut store = match self.store.lock() {
+                Ok(value) => value,
+                Err(err) => {
+                    log::warn!(
+                        "Could not acquire store lock. Error: {}",
+                        err
+                    );
+                    return;
+                }
+            };
+            let mut contents = if let Some(contents) = store.get(&key)
+            {
+                Cow::Borrowed(contents)
+            } else {
+                log::error!(
                 "textDocument/didChange called on unknown file {}",
                 key
             );
-            return;
+                return;
+            };
+            for change in params.content_changes {
+                contents =
+                    Cow::Owned(if let Some(range) = change.range {
+                        replace_string_in_range(
+                            contents.into_owned(),
+                            range,
+                            change.text,
+                        )
+                    } else {
+                        change.text
+                    });
+            }
+            let new_contents = contents.into_owned();
+            let c = new_contents.clone();
+            store.insert(key.clone(), new_contents);
+            c
         };
-        for change in params.content_changes {
-            contents =
-                Cow::Owned(if let Some(range) = change.range {
-                    replace_string_in_range(
-                        contents.into_owned(),
-                        range,
-                        change.text,
-                    )
-                } else {
-                    change.text
-                });
-        }
-        let new_contents = contents.into_owned();
-        store.insert(key.clone(), new_contents);
+        self.publish_diagnostics(&key, contents.as_str()).await;
```

Comment:
In the refactoring above, did you actually fix an issue, or was your intention just to limit the amount of time that the lock is retained, by scoping `let mut store = match self.store.lock()` to a smaller block?

Reply:
Two things changed:

1. I added the call to  `self.publish_diagnostics(&key, contents.as_str()).await;` Since this is an await call everything in the scope of that call must be `Send` because it may execute across multiple threads. However the `store` is not `Send`.
2. The change to add the new scoping blocks was to explicitly move the await call into its own block away from the `store` so that the `store` is dropped before we call await informing Rust that it doesn't need to be `Send`.